xmodule/services.py

```diff
@@ -17,13 +20,14 @@
 from common.djangoapps.track import contexts
 from lms.djangoapps.courseware.masquerade import is_masquerading_as_specific_student
 from xmodule.modulestore.django import modulestore
+from xmodule.graders import ShowCorrectness, ShowAnswer
 
 from lms.djangoapps.courseware.field_overrides import OverrideFieldData
 from lms.djangoapps.courseware.model_data import DjangoKeyValueStore, FieldDataCache
 from lms.djangoapps.lms_xblock.field_data import LmsFieldData
 from lms.djangoapps.lms_xblock.models import XBlockAsidesConfig
 
-from lms.djangoapps.grades.api import signals as grades_signals
+from lms.djangoapps.grades.signals import signals as grades_signals
```

Comment:
@Cup0fCoffee, why did we change this one? The idea is to import things from other apps through the `api` module.

Reply:
@Agrendalath I didn't dig too much into it (I'm going to dig a bit more after I post this comment), but I think it's because due to some `import` changes, a circular imports occur. I'm getting a following error when I change from `signals` to `api`:

```
Traceback (most recent call last):
  File "manage.py", line 103, in <module>
    startup.run()
  File "/edx/app/edxapp/edx-platform/lms/startup.py", line 20, in run
    django.setup()
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/__init__.py", line 24, in setup
    apps.populate(settings.INSTALLED_APPS)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/apps/registry.py", line 116, in populate
    app_config.import_models()
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/apps/config.py", line 269, in import_models
    self.models_module = import_module(models_module_name)
  File "/usr/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 848, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/bulk_email/models.py", line 22, in <module>
    from openedx.core.djangoapps.course_groups.cohorts import get_cohort_by_name
  File "/edx/app/edxapp/edx-platform/openedx/core/djangoapps/course_groups/cohorts.py", line 21, in <module>
    from lms.djangoapps.courseware import courses
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/courseware/courses.py", line 39, in <module>
    from lms.djangoapps.courseware.date_summary import (
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/courseware/date_summary.py", line 21, in <module>
    from lms.djangoapps.certificates.api import get_active_web_certificate, can_show_certificate_available_date_field
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/certificates/api.py", line 27, in <module>
    from lms.djangoapps.certificates.generation_handler import (
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/certificates/generation_handler.py", line 24, in <module>
    from lms.djangoapps.grades.api import CourseGradeFactory
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/grades/api.py", line 15, in <module>
    from lms.djangoapps.grades import constants, context, course_data, events
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/grades/context.py", line 10, in <module>
    from .course_grade import CourseGrade
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/grades/course_grade.py", line 17, in <module>
    from .subsection_grade import ZeroSubsectionGrade
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/grades/subsection_grade.py", line 16, in <module>
    from xmodule.services import ProblemFeedbackService
  File "/edx/app/edxapp/edx-platform/xmodule/services.py", line 30, in <module>
    from lms.djangoapps.grades.api import signals as grades_signals
ImportError: cannot import name 'signals' from partially initialized module 'lms.djangoapps.grades.api' (most likely due to a circular import) (/edx/app/edxapp/edx-platform/lms/djangoapps/grades/api.py)
```

Reply:
From the stack trace above, it can be observed that when importing from `grades.api`, it imports other modules from `grades`. The `subsection_grade` module tries to import from `xmodule.services`, which in turn tries to import from `grades.api`, which forms the circular dependency.

Reply:
In the `subsection_grade` the `ProblemFeedbackService`, which is imported from the `xmodule.services` is used only once in one of the methods. We can move the import from the top of the file into the method body, to avoid the circular dependency. That way the `xmodule.services` will be able to import from the `grades.api`. I'll push the commit now, but please let me know if you think that's not a good approach.